devicemodel/hw/pci/virtio/vhost_vsock.c

```diff
@@ -0,0 +1,340 @@
+		return -1;
+	}
+
+	vhost_vsock->vhost_started = true;
+	return 0;
+}
+
+static int
+vhost_vsock_stop(struct vhost_vsock *vhost_vsock)
+{
+	int rc;
+
+	if (!vhost_vsock->vhost_started) {
+		pr_err("vhost vsock is not started!\n");
+		return 0;
+	}
+
+	rc = vhost_vsock_set_running(&vhost_vsock->vdev, 0);
+	if (rc < 0) {
+		pr_err("vhost_vsock_set_running is failed with %d\n", rc);
+		return -1;
+	}
+
+	rc = vhost_dev_stop(&vhost_vsock->vdev);
+	if (rc < 0) {
+		pr_err("vhost_dev_stop is failed!\n");
+		return -1;
+	}
+
+	vhost_vsock->vhost_started = false;
+	return 0;
+}
+
+static void
+virtio_vsock_set_status(void *vdev, uint64_t status)
+{
+	struct virtio_vsock *vsock = vdev;
+	bool should_start = status & VIRTIO_CONFIG_S_DRIVER_OK;
+	int rc;
+
+	if (!vsock->vhost_vsock) {
+		pr_err("virtio_vsock_set_status vhost is NULL!\n");
+		return;
+	}
+
+	if (should_start) {
+		rc = vhost_vsock_start(vsock->vhost_vsock);
+		if (rc < 0) {
+			pr_err("vhost_vsock_start is failed!\n");
+			return;
+		}
+	} else if (vsock->vhost_vsock->vhost_started &&
+		should_start == 0) {
+		rc = vhost_vsock_stop(vsock->vhost_vsock);
+		if (rc < 0) {
+			pr_err("vhost_vsock_stop is failed!\n");
+			return;
+		}
+	}
+}
+
+static void
+virtio_vsock_apply_feature(void *vdev, uint64_t negotiated_features)
+{
+	struct virtio_vsock *vsock = vdev;
+
+	vsock->features = negotiated_features;
+}
+
+static int
+virtio_vsock_read_cfg(void *vdev, int offset, int size, uint32_t *retval)
+{
+	struct virtio_vsock *vsock = vdev;
+	void *ptr;
+
+	ptr = (uint8_t *)&vsock->config + offset;
+	memcpy(retval, ptr, size);
+
+	return 0;
+}
+
+static void
+virtio_vsock_reset(void *vdev)
+{
+	struct virtio_vsock *vsock = vdev;
+
+	pr_dbg(("vsock: device reset requested!\n"));
+	/* now reset rings, MSI-X vectors, and negotiated capabilities */
+	virtio_reset_dev(&vsock->base);
+}
+
+static struct virtio_ops virtio_vsock_ops = {
+	"vhost-vsock",			/* our name */
+	VHOST_VSOCK_MAXQ,		/* we currently support 2 virtqueues */
+	sizeof(struct virtio_vsock_config), /* config reg size */
+	virtio_vsock_reset,		/* reset */
+	NULL,				/* device-wide qnotify -- not used */
+	virtio_vsock_read_cfg,		/* read PCI config */
+	NULL,				/* write PCI config */
+	virtio_vsock_apply_feature,	/* apply negotiated features */
+	virtio_vsock_set_status,	/* called on guest set status */
+};
+
+static struct vhost_vsock *
+vhost_vsock_init(struct virtio_base *base, int vq_idx)
+{
+	struct vhost_vsock *vhost_vsock = NULL;
+	uint64_t vhost_features = VHOST_VSOCK_FEATURES;
+	int rc;
+
+	vhost_vsock = calloc(1, sizeof(struct vhost_vsock));
+	if (!vhost_vsock) {
+		pr_err(("vhost init out of memory!\n"));
+		goto fail;
+	}
+
+	/* pre-init before calling vhost_dev_init */
+	vhost_vsock->vdev.nvqs = 2;
+	vhost_vsock->vdev.vqs = vhost_vsock->vqs;
+	vhost_vsock->vhost_fd = open("/dev/vhost-vsock", O_RDWR);;
+	if (vhost_vsock->vhost_fd < 0) {
+		pr_err(("Open vhost-vsock fail, pls open vsock kernel config!\n"));
```

Comment:
Why the double parentheses?

The message does not make a lot of sense. I guess you means something like:
```
Failed to open the vhost vsock file descriptor, please make sure your kernel configuration has the CONFIG_VHOST_VSOCK turned on.
```